packages/network-controller/src/NetworkController.ts

```diff
@@ -353,42 +420,57 @@ export class NetworkController extends BaseControllerV2<
     this.#previousProviderConfig = this.state.providerConfig;
   }
 
-  #configureProvider(
-    type: NetworkType,
-    rpcUrl: string | undefined,
-    chainId: Hex | undefined,
-  ) {
-    switch (type) {
-      case NetworkType.mainnet:
-      case NetworkType.goerli:
-      case NetworkType.sepolia:
-        this.#setupInfuraProvider(type);
-        break;
-      case NetworkType.rpc:
-        if (chainId === undefined) {
-          throw new Error('chainId must be provided for custom RPC endpoints');
-        }
-
-        if (rpcUrl === undefined) {
-          throw new Error('rpcUrl must be provided for custom RPC endpoints');
-        }
-        this.#setupStandardProvider(rpcUrl, chainId);
-        break;
-      default:
-        throw new Error(`Unrecognized network type: '${type}'`);
-    }
-  }
-
+  /**
+   * Accesses the provider and block tracker for the currently selected network.
+   *
+   * @returns The proxy and block tracker proxies.
+   */
   getProviderAndBlockTracker(): {
-    provider: SwappableProxy<Provider> | undefined;
-    blockTracker: SwappableProxy<BlockTracker> | undefined;
+    provider: SwappableProxy<ProxyWithAccessibleTarget<Provider>> | undefined;
+    blockTracker:
+      | SwappableProxy<ProxyWithAccessibleTarget<BlockTracker>>
+      | undefined;
   } {
     return {
       provider: this.#providerProxy,
       blockTracker: this.#blockTrackerProxy,
     };
   }
 
+  /**
+   * Returns all of the network clients that have been created so far. This
+   * collection represents not only built-in networks but also any networks that
+   * the user has added.
+   *
+   * @returns The list of known network clients.
+   */
+  getNetworkClients(): AutoManagedNetworkClient<NetworkClientConfiguration>[] {
```

Comment:
Right. The first version of this method was called `getNetworkClientsById` and returned an object keyed by an identifier, which for Infura networks would look like `infura:<network>` and for custom networks would look like `custom:<chainId>:<lowercased rpcUrl>` [^fn1].

Ultimately I wasn't sure how we will end up using this collection and figured that once we had more information we could tweak this. So I went with an array for now. But, an object would definitely be more versatile. I'll play around with this some more.

[^fn1]: Custom networks aren't always associated with a network configuration ID, so we can't use that in the identifier (because the network controller can be instantiated with a `providerConfig` that has nothing to do with a built-in network or a network in `networkConfigurations`)

Reply:
>Custom networks aren't always associated with a network configuration ID

They are if you consider just how extension and mobile are using this controller. We can consider it a breaking change of this API, but "supporting custom networks specified via provider config that don't correspond with network configurations" is not a feature we're supporting intentionally. If anything it's a bug.

Reply:
Okay, I've changed this method to return an object rather than an array: [96c4229](https://github.com/MetaMask/core/pull/1439/commits/96c42295e5130ebbc32b199c43163e229f0e2d4e)